ENH: Added "seed" parameter to `subsample-ids`

VinzentRisch opened the pull request:
solves #320

- Adds seed parameter to `subsample-ids` function.
- Seed is per default set to 1. It can be set to "random" what creates a random 32 bit int as the seed. 
- Setting the seed to default 1 was done to ensure that the seed is logged in provenance. This improved reproducability.
- At this point it is not possible to have a random seed that is created in the function that will be recorded in provenance.

VinzentRisch committed:
added seed parameter with tests

VinzentRisch committed:
lint

gregcaporaso commented on q2_feature_table/_subsample_ids.py:
```diff
@@ -5,12 +5,14 @@
 
 def subsample_ids(table: biom.Table, subsampling_depth: int,
-                  axis: str) -> biom.Table:
+                  axis: str, seed: Union[int, str] = 1) -> biom.Table:
```
```suggestion
                  axis: str, seed: Union[int, None] = None) -> biom.Table:
```

gregcaporaso replied:
```suggestion
                  axis: str, seed: int = None) -> biom.Table:
```

gregcaporaso reviewed (changes requested):
@VinzentRisch, I thought I left a comment on this earlier today, but I see to have lost it so I'm thinking maybe it didn't save and I navigated away from the page. But, apologies if this feedback is showing up twice. 

I completely agree that we need to be storing random seeds in provenance, but the approach implemented here will be problematic. The vast majority of users won't notice or override this parameter setting, in which case the result will be that the same seed is used all the time. 

We have support for passing random seeds on the command line in q2-boots - see how we handle this here:
https://github.com/caporaso-lab/q2-boots/commit/9195f9cfac1694196e1dfcc68934301d9238b350

Could you update this PR to use this approach? 

I created a framework issue to develop an approach for better supporting this:
https://github.com/qiime2/qiime2/issues/823

gregcaporaso commented on q2_feature_table/_subsample_ids.py:
```diff
@@ -20,8 +22,14 @@ def subsample_ids(table: biom.Table, subsampling_depth: int,
                          'is: %d.' % len(table.ids()))
 
+    # Generate a random seed if seed is None
+    if seed == "random":
```
You can use the built-in [`random.seed` function](https://docs.python.org/3/library/random.html#random.seed) for this.

gregcaporaso commented:
Oh, I see - my previous comment was on #321. This feedback is relevant to both of the PRs.

gregcaporaso commented on q2_feature_table/_subsample_ids.py:
```diff
@@ -20,8 +22,14 @@ def subsample_ids(table: biom.Table, subsampling_depth: int,
                          'elements on the desired axis. The maximum depth '
                          'is: %d.' % len(table.ids()))
 
+    # Generate a random seed if seed is None
+    if seed == "random":
+        rng = np.random.default_rng()
+        seed = rng.integers(0, 2 ** 32 - 1)
+
```
This should all be able to go away since this happens in `Table.subsample` - right? 

```suggestion
```

gregcaporaso commented on q2_feature_table/plugin_setup.py:
```diff
@@ -61,7 +61,9 @@
     parameters={'subsampling_depth': Int % Range(1, None),
-                'axis': Str % Choices(['sample', 'feature'])},
+                'axis': Str % Choices(['sample', 'feature']),
+                'seed': Int % Range(0, 2**32) | Str % Choices(["random"])
```
```suggestion
                'seed': Int % Range(0, None)
```
```suggestion
                'seed': Int % Range(0, 2**32) | Str % Choices(["random"])
```

gregcaporaso commented on q2_feature_table/plugin_setup.py:
```diff
@@ -71,7 +73,11 @@
                  'randomly selected to be retained. If "feature" then '
-                 'a random set of features will be selected to be retained.')
+                 'a random set of features will be selected to be retained.'),
+        'seed': ('Set the seed for the subsampling. Using the same seed with '
+                 'the same table will always lead to the same result. Using '
+                 '"random", sets the seed to a random number. The random '
+                 'seed will not be logged in provenance.')
```
Edits to the text. I'd leave the bit about provenance out, since we'll ideally be fixing that very shortly and it's common to all actions that take seeds right now. 

```suggestion
        'seed': ('Set the seed for the subsampling. Using the same seed with '
                 'the same table will always lead to the same result. By '
                 'default a random seed will be used.')
```

gregcaporaso commented on q2_feature_table/tests/test_subsample.py:
```diff
@@ -21,7 +21,7 @@ def test_subsample_samples(self):
                   ['O1', 'O2'],
                   ['S1', 'S2', 'S3'])
-        a = subsample_ids(t, 2, 'sample')
+        a = subsample_ids(t, 2, 'sample', 'random')
```
Can you adapt these tests to be similar to the ones we put together for #321?

gregcaporaso reviewed (changes requested):
@VinzentRisch, could you update this PR to mirror the approaches we've taken to this functionality and the tests in #321?

VinzentRisch committed:
merge dev

gregcaporaso commented:
@VinzentRisch - is this still something you're interested in working on?

VinzentRisch commented:
Hey @gregcaporaso 
Sorry! this got lost on my todo list. I will take a look today.

VinzentRisch committed:
Merge branch 'dev' into 55_seed

VinzentRisch committed:
added tests with iteration

VinzentRisch committed:
lint

VinzentRisch commented:
Hey @gregcaporaso 
Again sorry for the oversight. I added all the changes similar to https://github.com/qiime2/q2-feature-table/pull/321.

gregcaporaso committed:
MAINT: swap error messages

these were backwards, and read as if they're printed on success, but they're printed on failure

gregcaporaso reviewed (approved):
@VinzentRisch, this looks good, thank you! I made a fix (and a simplification) to the tests [here](https://github.com/qiime2/q2-feature-table/pull/317/commits/939dde3930507f51478fdea486259f20346df28f). The tests were all right, but the message that would have been printed on failure (regarding the use or non-use of a random seed) was backwards such that a confusing message would have been printed on failure. 

This is ready for merge once the tests pass.
